### PR TITLE
[ios][core] Automatically register expo modules

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

Since we merged `@unimodules/core` and `@unimodules/react-native-adapter`, it's now possible to remove a bunch of setup steps. We can automatically register "extra modules" instead of asking users to do this in the bridge delegate (usually `AppDelegate`).

# How

Instead of registering our bridge modules through `extraModulesForBridge:`, we can register one of our modules (`EXNativeModulesProxy`) using `RCT_EXPORT_MODULE` macro and once it gets a bridge use `registerAdditionalModuleClasses:` to register the others.
There is one subtle difference in these two approaches — the latter method expects module classes instead of module instances and that's mostly why I had to change many files.

With these changes, expo modules setup on iOS is reduced to autolinking and extending `AppDelegate` with `EXAppDelegateWrapper`. `EXModuleRegistryAdapter` and `EXModuleRegistryProvider` are deprecated now.

# Test Plan

- Tested two cases in bare-expo
  - with the old setup — we get deprecation warnings and everything works as previously
  - with no setup — everything works as expected
- Expo Go seems to work as expected, after a few modifications in `EXVersionManager`

### CODE REVIEW

File: packages/expo-modules-core/ios/ModuleRegistryAdapter/EXModuleRegistryAdapter.m [hunk 1]
```diff
@@ -70,6 +73,7 @@ - (instancetype)initWithModuleRegistryProvider:(EXModuleRegistryProvider *)modul
   // subclass EXViewManagerAdapter, so RN expects to find EXViewManagerAdapter
   // exported.
   [extraModules addObject:[[EXViewManagerAdapter alloc] init]];
+  [extraModules addObject:[[ViewModuleWrapper alloc] initWithDummy:nil]];
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/ModuleRegistryAdapter/EXModuleRegistryAdapter.m:76]: `ViewModuleWrapper` is a base class of our swifty view managers, but `init` with no arguments would throw fatal exception, so I omit this by another convenience init.

File: packages/expo-modules-core/ios/ModuleRegistryAdapter/EXViewManagerAdapterClassesRegistry.h [hunk 2]
```diff
@@ -12,5 +12,6 @@
 @interface EXViewManagerAdapterClassesRegistry : NSObject
 
 - (Class)viewManagerAdapterClassForViewManager:(EXViewManager *)viewManager;
++ (Class)createViewManagerAdapterClassForViewManager:(EXViewManager *)viewManager;
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/ModuleRegistryAdapter/EXViewManagerAdapterClassesRegistry.h:15]: This is almost the same as the method above, but static and doesn't cache results.

File: packages/expo-modules-core/react-native.config.js [hunk 3]
```diff
@@ -1,6 +1,12 @@
+const path = require('path');
+
 module.exports = {
   dependency: {
     platforms: {
+      ios: {
+        podspecPath: path.join(__dirname, 'ios/ExpoModulesCore.podspec'),
+        project: 'ios/ExpoModulesCore.xcodeproj',
+      },
```
Comment (tsapeta) [inline packages/expo-modules-core/react-native.config.js:9]: This makes `expo-modules-core` linkable by `react-native` CLI. Don't ask me why `podspecPath` must be absolute whereas `project` must be relative 👀 This `*.xcodeproj` doesn't really exist, but passing something here is required to make it linkable 🤷‍♂️

Comment (brentvatne) [reply packages/expo-modules-core/react-native.config.js:9]: i guess this is something that will change when moving to expo package

**Review by brentvatne — APPROVED** [review]

File: ios/Exponent/Versioned/Core/EXVersionManager.mm [hunk 4]
```diff
@@ -364,7 +364,11 @@ - (NSArray *)extraModulesForBridge:(id)bridge
   id<EXModuleRegistryDelegate> moduleRegistryDelegate = [[resolverClass alloc] initWithParams:params];
   [moduleRegistryProvider setModuleRegistryDelegate:moduleRegistryDelegate];
 
-  EXScopedModuleRegistryAdapter *moduleRegistryAdapter = [[EXScopedModuleRegistryAdapter alloc] initWithModuleRegistryProvider:moduleRegistryProvider swiftModulesProviderClass:NSClassFromString(@"ExpoModulesProvider")];
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+  EXScopedModuleRegistryAdapter *moduleRegistryAdapter = [[EXScopedModuleRegistryAdapter alloc] initWithModuleRegistryProvider:moduleRegistryProvider];
+#pragma clang diagnostic pop
+
```
[the remark this answers is not in the document]
Comment (tsapeta) [reply ios/Exponent/Versioned/Core/EXVersionManager.mm:371]: Yes, I have a plan for it 🙂 I think we don't need to scope (extend) module registry adapter anymore, but this needs a bit of refactor that I don't wanna do right now.

**Pull request merged**